stdlib/src/memory/boxed.mojo

```diff
@@ -0,0 +1,112 @@
+# ===----------------------------------------------------------------------=== #
+# Copyright (c) 2024, Modular Inc. All rights reserved.
+#
+# Licensed under the Apache License v2.0 with LLVM Exceptions:
+# https://llvm.org/LICENSE.txt
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ===----------------------------------------------------------------------=== #
+from memory import UnsafePointer, stack_allocation, memcpy
+
+struct Boxed[T: AnyType, address_space: AddressSpace = AddressSpace.GENERIC]:
+    """A safe, owning, smart pointer.
+
+    This smart pointer is designed for cases where there is clear ownership
+    of the underlying data, and restricts access to it through the lifetime
+    system such that no more than one mutable alias for the underlying data
+    may exist. Consider Boxed[T] over UnsafePointer[T] where possible.
+    """
+
+    var _inner: UnsafePointer[T, address_space]
+
+    fn __init__[T: Movable](inout self: Boxed[T, AddressSpace.GENERIC], owned value: T):
+        """Construct a new Boxed[] by moving the passed value into a new backing allocation.
+
+        Parameters:
+            T: The type of the data to store. It is restricted to `Movable` here to allow efficient move construction.
+
+        Args:
+            value: The value to move into the Boxed[].
+        """
+        self._inner = UnsafePointer[T, AddressSpace.GENERIC].alloc(1)
+        self._inner.init_pointee_move(value^)
+
+    fn __init__[T: ExplicitlyCopyable](inout self: Boxed[T, AddressSpace.GENERIC], *, copy_value: T):
+        """Construct a new Boxed[] by explicitly copying the passed value into a new backing allocation.
+
+        Parameters:
+            T: The type of the data to store.
+
+        Args:
+            copy_value: The value to explicitly copy into the Boxed[].
+        """
+        self._inner = UnsafePointer[T, address_space].alloc(1)
+        self._inner.init_pointee_explicit_copy(copy_value)
+
+    fn __init__[T: ExplicitlyCopyable](inout self: Boxed[T, AddressSpace.GENERIC], *, copy_box: Boxed[T, AddressSpace.GENERIC]):
```

Comment:
these are way over 80 characters, you can setup a ruler in your IDE for the docstrings, but the code should be auto formatted by mojo format when saving a file (If you set it up)

Reply:
I just need to spend a few minutes fixing the hook again \:)